### Differential transcript usage unravels transcriptional alterations in Alzheimer's disease human brains

#### Diego Marques-Coelho<sup>1,2</sup>, Lukas da Cruz Carvalho Iohan<sup>1,2</sup>,  Ana Raquel Melo de Farias<sup>3</sup>, Amandine Flaig<sup>3</sup>, The Brainbank Neuro–CEB Neuropathology Network, Jean-Charles Lambert<sup>3</sup>, Marcos Romualdo Costa<sup>1,3</sup>

1 Brain Institute, Federal University of Rio Grande do Norte, Av. Nascimento de Castro 2155 Natal, Brazil
2 Bioinformatics Multidisciplinary Environment (BioME), Federal University of Rio Grande do Norte
3 Unité INSERM 1167, RID-AGE-Risk Factors and Molecular Determinants of Aging-Related Diseases, Institut Pasteur de Lille, University of Lille, U1167-Excellence Laboratory LabEx DISTALZ, BP 245, 1 rue du Professeur Calmette, 59019 Lille Cedex, France

* Corresponding author: marcos.costa@pasteur-lille.fr or
* Corresponding author: diegomscoelho@gmail.com

Publication link: https://www.nature.com/articles/s41514-020-00052-5

## Abstract
Alzheimer’s disease (AD) is the leading cause of dementia in aging individuals. Yet, the pathophysiological processes involved in AD onset and progression are still poorly understood. Among numerous strategies, a comprehensive overview of gene expression alterations in the diseased brain could contribute for a better understanding of the AD pathology. In this work, we probed the differential expression of genes in different brain regions of healthy and AD adult subjects using data from three large transcriptomic studies: Mayo Clinic, Mount Sinai Brain Bank (MSBB), and ROSMAP. Using a combination of differential expression of gene and isoform switch analyses, we provide a detailed landscape of gene expression alterations in the temporal and frontal lobes, harboring brain areas affected at early and late stages of the AD pathology, respectively. Next, we took advantage of an indirect approach to assign the complex gene expression changes revealed in bulk RNAseq to individual cell types/subtypes of the adult brain. This strategy allowed us to identify previously overlooked gene expression changes in the brain of AD patients. Among these alterations, we show isoform switches in the AD causal gene amyloid-beta precursor protein (APP) and the risk gene bridging integrator 1 (BIN1), which could have important functional consequences in neuronal cells. Altogether, our work proposes a novel integrative strategy to analyze RNAseq data in AD and other neurodegenerative diseases based on both gene/transcript expression and regional/cell-type specificities.
